### PR TITLE
[FLINK-11701][table-runtime-blink] Introduce an abstract set of data formats

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Blink uses an abstract set of data formats to make internal calculations use the binary format as much as possible. This minimizes the serialization overhead and java object overhead.

## Brief change log

Data format corresponding relation:
- BaseRow(BinaryRow and GenericRow) <=> Row
- int <=> Date //Flink used to use int in the calculation, but the remaining in Row is still Date, we will change it completely.
- int <=> Time
- long <=> Timestamp
- byte[] <=> byte[]
- primitive type keep same, but use less boxed type.

We decided to build BinaryRow on multiple Segments because we wanted fine-grained memory control. For fast access, we keep the primitive type on a single Segment.

## Verifying this change
Unit test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by JingsongLi** [comment]
After https://github.com/apache/flink/pull/7790 merge

**Comment by KurtYoung** [comment]
Could you also change the commit message and the title after moving to runtime?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+package org.apache.flink.table.util;
+
+import org.apache.flink.core.memory.MemorySegment;
+
+import java.nio.ByteOrder;
+
+/**
+ * Util for data format segments calc.
+ */
+public class SegmentsUtil {
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java]: Could you add some comments to each util method this class provided, just like MemorySegment did?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryFormat.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	 * <p>If length is greater or equal to 8, its binary format is:
+	 * 4bytes variable part offset and 4bytes length. Data is stored in variable-length part.
+	 *
+	 * <p>Note: Need to consider the ByteOrder.
+	 *
+	 * @param baseOffset base offset of composite binary format.
+	 * @param fieldOffset absolute start offset of 'variablePartOffsetAndSize'.
+	 * @param variablePartOffsetAndSize a long value, 4bytes variable part offset and 4bytes length.
+	 */
+	static BinaryString readBinaryStringFieldFromSegments(
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryFormat.java]: Seems this method should place in `BinaryString`?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryFormat.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	public boolean binaryEquals(BinaryFormat that) {
+		return sizeInBytes == that.sizeInBytes &&
+				SegmentsUtil.equals(segments, offset, that.segments, that.offset, sizeInBytes);
+	}
+
+	/**
+	 * Get binary string, if length less than 8, will be include in variablePartOffsetAndSize.
+	 *
+	 * <p>If length is less than 8, its binary format is:
+	 * 1bit mark(1), 7bits length, and 7bytes data.
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryFormat.java]: Could you explain what is mark for?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+/**
+ * A special row which is backed by {@link MemorySegment} instead of Object. It can significantly reduce the
+ * serialization/deserialization of Java objects.
+ *
+ * <p>A Row has two part: Fixed-length part and variable-length part.
+ *
+ * <p>Fixed-length part contains null bit set and field values. Null bit set is used for null tracking and is
+ * aligned to 8-byte word boundaries. `Field values` holds fixed-length primitive types and variable-length
+ * values which can be stored in 8 bytes inside. If it do not fit the variable-length field, then store the
+ * length and offset of variable-length part. Fixed-length part will certainly fall into a MemorySegment,
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java]: Add some more explains about the fixed part behavior, like during the write phase, if the target memory segment has less space than fixed length part size, we will skip the space

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * A special row which is backed by {@link MemorySegment} instead of Object. It can significantly reduce the
+ * serialization/deserialization of Java objects.
+ *
+ * <p>A Row has two part: Fixed-length part and variable-length part.
+ *
+ * <p>Fixed-length part contains null bit set and field values. Null bit set is used for null tracking and is
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java]: Header also in fixed length part

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+		int fieldOffset = getFieldOffset(pos);
+		final long offsetAndSize = segments[0].getLong(fieldOffset);
+		return BinaryFormat.readBinaryStringFieldFromSegments(segments, offset, fieldOffset, offsetAndSize);
+	}
+
+	/**
+	 * The bit is 1 when the field is null. Default is 0.
+	 */
+	public boolean anyNull() {
+		for (int i = 0; i < nullBitsSizeInBytes; i += 8) {
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java]: Looks like you don't skip the header

Comment (KurtYoung) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java]: Add a test case for this

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryStringUtil.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.dataformat;
+
+/**
+ * Util for {@link BinaryString}.
+ */
+public class BinaryStringUtil {
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryStringUtil.java]: Put this class to `util` package?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryStringUtil.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+	private static final int MAX_BYTES_LENGTH = 1024 * 64;
+	private static final ThreadLocal<byte[]> BYTES_LOCAL = new ThreadLocal<>();
+
+	public static byte[] allocateBytes(int length) {
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryStringUtil.java]: I think you should remind people that this byte[] is only for temporary usage, it should not be stored in somewhere else. Also, you can explain what the thread local cache is used for, what kind of scenarios you want to improve here.

**Comment by JingsongLi** [comment]
Fix comments

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+	}
+
+	public void setTotalSize(int sizeInBytes) {
+		this.sizeInBytes = sizeInBytes;
+	}
+
+	private void setNotNullAt(int i) {
+		assertIndexIsValid(i);
+		// need add header 8 bit.
+		SegmentsUtil.bitUnSet(segments[0], offset, i + 8);
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java]: How about we turn the magic number 8 to something like HEAER_LENGTH ?

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java]: OK

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryWriter.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+ *
+ * <p>If write a format binary:
+ * 1. New a writer.
+ * 2. Write each field by writeXX or setNullAt. (Variable length fields can not be written
+ * repeatedly.)
+ * 3. Invoke {@link #complete()}.
+ *
+ * <p>If want to reuse this writer, please invoke {@link #reset()} first.
+ */
+public abstract class BinaryWriter {
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryWriter.java]: I think we need a better abstraction for `BinaryWriter`. There are 2 issues from my point:
1. We don't have a clear interface for `BinaryWriter`, maybe we should make this `BinaryWriter` an interface, and move some common implementation to `AbstractBinaryWriter`
2. Some implementation here seems like BinaryRow specific, like `writeBytesToFixLenPart`, we shouldn't know what is fix length part and what is variable length part here.

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryWriter.java]: BinaryArray has fix length part too.
The difference is only the fix length part of BinaryRow must be in a single MemorySegment.

**Comment by KurtYoung** [comment]
merging...

**Pull request closed without merging**